developer/src/kmc-keyboard/test/test-layr.ts

```diff
@@ -90,4 +87,30 @@ describe('layr', function () {
     assert.equal(touch0row0.keys.length, 4);
     allKeysOk(touch0row0,'Q q W w', 'touch0row0');
   });
+  testCompilationCases(LayrCompiler, compilerTestCallbacks, [
+    {
+      subpath: 'sections/layr/invalid-invalid-hardware.xml',
+      errors: [CompilerMessages.Error_InvalidHardware({hardware: 'stenography'})],
+    },
+    {
+      subpath: 'sections/layr/invalid-missing-hardware.xml',
+      errors: [CompilerMessages.Error_MissingHardware()],
+    },
+    {
+      subpath: 'sections/layr/invalid-multi-hardware.xml',
+      errors: [CompilerMessages.Error_MustHaveAtMostOneLayersElementPerForm({ form: 'hardware' })],
+    },
+    {
+      subpath: 'sections/layr/invalid-touch-hardware.xml',
+      errors: [CompilerMessages.Error_NoHardwareOnTouch({ hardware: 'iso' })],
```

Comment:
@mcdurdin in tests that have lots of copypasta I've found it's very easy to make mistakes, i.e. miss actual assertions.. A framework that handles the boilerplate cuts down on on that, and makes future refactorings easier.